pkg/catalog/manager/manager.go

```diff
@@ -184,3 +194,108 @@ func (m *Manager) deleteBadCatalogTemplates() []error {
 
 	return errs
 }
+
+func (m *Manager) ValidateChartCompatibility(template *v3.CatalogTemplateVersion, clusterName string) error {
+	if err := m.ValidateRancherVersion(template); err != nil {
+		return err
+	}
+	if err := m.ValidateKubeVersion(template, clusterName); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Manager) ValidateKubeVersion(template *v3.CatalogTemplateVersion, clusterName string) error {
+	if template.Spec.KubeVersion == "" {
+		return nil
+	}
+	constraint, err := semver.ParseRange(template.Spec.KubeVersion)
+	if err != nil {
+		logrus.Errorf("failed to parse constraint for kubeversion %s: %v", template.Spec.KubeVersion, err)
+		return nil
+	}
+
+	cluster, err := m.clusterLister.Get("", clusterName)
+	if err != nil {
+		return err
+	}
+
+	k8sVersion, err := semver.Parse(strings.TrimPrefix(cluster.Status.Version.String(), "v"))
+	if err != nil {
+		return err
+	}
+	if !constraint(k8sVersion) {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("incompatible kubernetes version [%s] for template [%s]", k8sVersion.String(), template.Name))
```

Comment:
None of the functions in here should return an APIerror, these helpers are getting called from all over and the caller should decide if they need to wrap as an APIerror